### PR TITLE
SW-801 Export variants with different attribute values v1

### PR DESCRIPTION
**Pull request by bojand-soprex**

### CODE REVIEW

**Review by TobiasGraml11 — CHANGES_REQUESTED** [review]

File: src/Export/Adapters/ExportItemAdapter.php [hunk 1]
```diff
@@ -24,15 +25,18 @@ class ExportItemAdapter
 
     private LoggerInterface $logger;
 
+    protected PluginConfig $pluginConfig;
```
Comment (nenadj-soprex) [inline src/Export/Adapters/ExportItemAdapter.php:28]: Add empty line here please

File: tests/Export/Adapters/ExportItemAdapterTest.php [hunk 2]
```diff
@@ -26,6 +27,12 @@ class ExportItemAdapterTest extends TestCase
     use ProductHelper;
     use ServicesHelper;
 
+    public PluginConfig $pluginConfig;
+
+    public function setUp(): void
+    {
+        $this->pluginConfig = $this->getPluginConfig();
+    }
```
Comment (nenadj-soprex) [inline tests/Export/Adapters/ExportItemAdapterTest.php:35]: Also add one empty row between methods

File: src/Export/Adapters/AdapterFactory.php [hunk 3]
```diff
@@ -72,6 +77,7 @@ public function __construct(
         $this->shopwarePropertiesAdapter = $shopwarePropertiesAdapter;
         $this->urlAdapter = $urlAdapter;
         $this->userGroupsAdapter = $userGroupsAdapter;
+        $this->variantConfigurationAdapter = $variantConfigurationAdapter;
```
Comment (TobiasGraml11) [inline src/Export/Adapters/AdapterFactory.php:80]: The option adapter is not loaded within the constructor

File: src/Export/Adapters/AttributeAdapter.php [hunk 4]
```diff
@@ -51,15 +51,13 @@ public function adapt(ProductEntity $product): array
         $categoryAttributes = $this->getCategoryAndCatUrlAttributes($product);
         $manufacturerAttributes = $this->getManufacturerAttributes($product);
         $propertyAttributes = $this->getPropertyAttributes($product);
-        $optionAttributes = $this->getOptionAttributes($product);
```
Comment (TobiasGraml11) [inline src/Export/Adapters/AttributeAdapter.php:54]: Keep this for extension compatibility.

File: src/Export/Adapters/ExportItemAdapter.php [hunk 5]
```diff
@@ -142,16 +147,27 @@ public function adaptProduct(Item $item, ProductEntity $product): ?Item
...
+            if ($this->pluginConfig->getMainVariant() != 'default') {
+                foreach ($this->adapterFactory->getOptionsAdapter()->getOptionAttributes($product) as $attribute) {
+                    $item->addMergedAttribute($attribute);
+                }
+            } else {
+                $attributes = $this->adapterFactory->getVariantConfigurationAdapter()
+                    ->getOptionAttributes($product);
+                foreach ($attributes as $attribute) {
+                    $item->addMergedAttribute($attribute);
+                }
```
Comment (TobiasGraml11) [inline src/Export/Adapters/ExportItemAdapter.php:170]: The initial idea was to just call the `optionsAdapter` within the `adaptProduct` function and the `variantConfigurationAdapter` within the `adaptVariant` function.
There should be no config check needed within this class.

File: src/Export/Adapters/OptionsAdapter.php [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+use FINDOLOGIC\Export\Data\Attribute;
+use FINDOLOGIC\Shopware6Common\Export\Config\PluginConfig;
+use FINDOLOGIC\Shopware6Common\Export\Utils\Utils;
+use Vin\ShopwareSdk\Data\Entity\Product\ProductEntity;
+use Vin\ShopwareSdk\Data\Entity\PropertyGroupOption\PropertyGroupOptionCollection;
+use Vin\ShopwareSdk\Data\Entity\PropertyGroupOption\PropertyGroupOptionEntity;
+
+class OptionsAdapter
+{
+
```
Comment (TobiasGraml11) [inline src/Export/Adapters/OptionsAdapter.php:16]:
```suggestion
```

File: src/Export/Adapters/OptionsAdapter.php [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    /**
+     * @return Attribute[]
+     */
+    public function getOptionAttributes(ProductEntity $product): array
+    {
+        if (!$product->options || !$product->options->count()) {
+            return [];
+        }
+
+        return Utils::getPropertyGroupOptionAttributes($product->options, $this->pluginConfig);
```
Comment (TobiasGraml11) [inline src/Export/Adapters/OptionsAdapter.php:34]: I know I suggested this, but I just realized that a Trait is the only solution. This way, the functions can not be extended within a extension.
With a trait, you can also remove the functions from the `AttributeAdapter`, as there should not be an issue with existing extensions.

**Review by TobiasGraml11 — CHANGES_REQUESTED** [review]

File: tests/Export/Adapters/AttributeAdapterTest.php [hunk 8]
```diff
@@ -345,7 +345,7 @@ public function testEmptyCategoryNameShouldStillExportCategory(): void
 
         $attributes = $this->attributeAdapter->adapt($productEntity);
 
-        $this->assertCount(7, $attributes);
+        $this->assertCount(5, $attributes);
```
Comment (TobiasGraml11) [inline tests/Export/Adapters/AttributeAdapterTest.php:348]: We'll need a `OptionsAdapterTest` for the 2 attributes

File: tests/Export/Adapters/AttributeAdapterTest.php [hunk 9]
```diff
@@ -409,7 +409,6 @@ public function testOnlyUniqueCategoriesAreExported(bool $isParentAssigned, bool
         $exportItemAdapter = $this->getExportItemAdapter(null, $config);
 
         $item = $exportItemAdapter->adapt($initialItem, $productEntity);
-
```
Comment (TobiasGraml11) [inline tests/Export/Adapters/AttributeAdapterTest.php:412]: Please keep the empty lines in the file

File: tests/Export/Adapters/ExportItemAdapterTest.php [hunk 10]
```diff
@@ -39,6 +47,12 @@ public function testEventsAreDispatched(): void
         /** @var EventDispatcher $eventDispatcherMock */
         $adapter = $this->getExportItemAdapter(null, null, null, $eventDispatcherMock);
         $adapter->adapt($xmlItem, $product);
+        $id = Uuid::randomHex();
```
Comment (TobiasGraml11) [inline tests/Export/Adapters/ExportItemAdapterTest.php:50]: Why is this change needed? There is no change regarding event dispatching.

File: tests/Traits/AttributeHelper.php [hunk 11]
```diff
@@ -47,13 +47,6 @@ public function getAttributes(ProductEntity $productEntity, string $integrationT
             ],
         );
 
-        foreach ($productEntity->options ?? [] as $option) {
```
Comment (TobiasGraml11) [inline tests/Traits/AttributeHelper.php:50]: This must be moved and tested somewhere else, instead of completely removed

**Review by TobiasGraml11 — CHANGES_REQUESTED** [review]

File: src/Export/Adapters/AdapterFactory.php [hunk 12]
```diff
@@ -54,7 +58,9 @@ public function __construct(
         SummaryAdapter $summaryAdapter,
         ShopwarePropertiesAdapter $shopwarePropertiesAdapter,
         UrlAdapter $urlAdapter,
-        UserGroupsAdapter $userGroupsAdapter
+        UserGroupsAdapter $userGroupsAdapter,
+        VariantConfigurationAdapter $variantConfigurationAdapter,
+        OptionsAdapter $optionsAdapter,
```
Comment (TobiasGraml11) [inline src/Export/Adapters/AdapterFactory.php:63]: Please keep the alphabetic order

File: src/Export/Adapters/AttributeAdapter.php [hunk 13]
```diff
@@ -148,69 +149,25 @@ protected function getManufacturerAttributes(ProductEntity $product): array
         return $attributes;
     }
 
+    //@deprecated tag:6.0.0 - Logic was moved to the OptionsAdapter
```
Comment (TobiasGraml11) [inline src/Export/Adapters/AttributeAdapter.php:152]: Just add it to the existing docblock

File: src/Traits/AdapterHelper.php [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+        return $attributes;
+    }
+
+    /**
+     * @return Attribute[]
+     */
+    protected static function getAttributePropertyAsAttribute(
+        PropertyGroupOptionEntity $propertyGroupOptionEntity,
+        PluginConfig $pluginConfig,
```
Comment (TobiasGraml11) [inline src/Traits/AdapterHelper.php:44]: You can also use member variables within traits, no need to provide them as parameter

File: src/Export/Adapters/ExportItemAdapter.php [hunk 15]
```diff
@@ -136,27 +141,37 @@ public function adaptProduct(Item $item, ProductEntity $product): ?Item
...
+                BeforeVariantAdaptEvent::NAME,
+            );
         }
 
         try {
             foreach ($this->adapterFactory->getOrderNumbersAdapter()->adapt($product) as $orderNumber) {
                 $item->addOrdernumber($orderNumber);
             }
 
-            foreach ($this->adapterFactory->getAttributeAdapter()->adapt($product) as $attribute) {
```
Comment (TobiasGraml11) [inline src/Export/Adapters/ExportItemAdapter.php:153]: The attributes adapter will still be needed, the issue is just about the options.

File: src/Export/Adapters/ExportItemAdapter.php [hunk 16]
```diff
@@ -136,27 +141,37 @@ public function adaptProduct(Item $item, ProductEntity $product): ?Item
...
             foreach ($this->adapterFactory->getShopwarePropertiesAdapter()->adapt($product) as $property) {
                 $item->addProperty($property);
             }
+
+            $attributes = $this->adapterFactory->getVariantConfigurationAdapter()
+                ->getOptionAttributes($product);
+
+            foreach ($attributes as $attribute) {
+                $item->addMergedAttribute($attribute);
+            }
```
Comment (TobiasGraml11) [inline src/Export/Adapters/ExportItemAdapter.php:174]:
```suggestion
            foreach ($this->adapterFactory->getVariantConfigurationAdapter()
                ->getOptionAttributes($product) as $attribute) {
                $item->addMergedAttribute($attribute);
            }
```

File: src/Export/Adapters/VariantConfigurationAdapter.php [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    /**
+     * @return Attribute[]
+     */
+    public function getOptionAttributes(ProductEntity $product): array
+    {
+        $options = $product->options;
+
+        $isVariant = !is_null($product->parentId);
```
Comment (TobiasGraml11) [inline src/Export/Adapters/VariantConfigurationAdapter.php:32]: This adapter is only called for variants, there is no check needed anymore.

**Review by TobiasGraml11 — CHANGES_REQUESTED** [review]

File: src/Traits/AdapterHelper.php [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+use FINDOLOGIC\Shopware6Common\Export\Utils\Utils;
+use Vin\ShopwareSdk\Data\Entity\PropertyGroupOption\PropertyGroupOptionCollection;
+use Vin\ShopwareSdk\Data\Entity\PropertyGroupOption\PropertyGroupOptionEntity;
+
+trait AdapterHelper
+{
+    /**
+     * @return Attribute[]
+     */
+    public static function getPropertyGroupOptionAttributes(
```
Comment (TobiasGraml11) [inline src/Traits/AdapterHelper.php:18]: Keep the functions protected and non-static

File: src/Export/Adapters/ExportItemAdapter.php [hunk 19]
```diff
@@ -162,14 +162,16 @@ public function adaptVariant(Item $item, ProductEntity $product): ?Item
...
             foreach ($this->adapterFactory->getShopwarePropertiesAdapter()->adapt($product) as $property) {
                 $item->addProperty($property);
             }
 
-            $attributes = $this->adapterFactory->getVariantConfigurationAdapter()
-                ->getOptionAttributes($product);
-
-            foreach ($attributes as $attribute) {
+            foreach ($this->adapterFactory->getVariantConfigurationAdapter()
+                         ->getOptionAttributes($product) as $attribute) {
```
Comment (TobiasGraml11) [inline src/Export/Adapters/ExportItemAdapter.php:174]: If this is reaching the max line length, please keep it in the variable, because this looks very weird 😅 

Or maybe try it like this:
```
foreach(
    ...
) {
}
```

File: src/Export/Adapters/VariantConfigurationAdapter.php [hunk 20]
```diff
@@ -29,30 +29,26 @@ public function getOptionAttributes(ProductEntity $product): array
...
+            function (array $listing) {
+                return $listing['expressionForListings'];
+            },
+        );
+        $variantListingGroupId = array_map(fn ($listing) => $listing['id'], $variantlisting);
+
+        if (
+            $this->pluginConfig->getMainVariant() === MainVariant::SHOPWARE_DEFAULT &&
+            count($variantlisting) &&
+            !$product->variantListingConfig['displayParent']
```
Comment (TobiasGraml11) [inline src/Export/Adapters/VariantConfigurationAdapter.php:47]: Aren't we ignoring the `displayParent` configuration for the general export with Shopware Default anyway?
So if we fetch the products while ignoring this config, then we can also do it here for consistency.

File: src/Traits/AdapterHelper.php [hunk 21]
```diff
@@ -12,25 +12,25 @@
 
 trait AdapterHelper
 {
+    protected PropertyGroupOptionEntity $propertyGroupOptionEntity;
+    protected PluginConfig $pluginConfig;
```
Comment (TobiasGraml11) [inline src/Traits/AdapterHelper.php:16]:
```suggestion
    protected PropertyGroupOptionEntity $propertyGroupOptionEntity;

    protected PluginConfig $pluginConfig;
```

File: src/Traits/AdapterHelper.php [hunk 22]
```diff
@@ -12,25 +12,25 @@
...
     /**
      * @return Attribute[]
      */
-    public static function getPropertyGroupOptionAttributes(
+    protected function getPropertyGroupOptionAttributes(
         PropertyGroupOptionCollection $collection,
-        PluginConfig $pluginConfig,
+        PluginConfig $pluginConfig
     ): array {
+        $this->pluginConfig = $pluginConfig;
```
Comment (TobiasGraml11) [inline src/Traits/AdapterHelper.php:25]: The plugin config is already set within the constructor of both adapters, so you can remove it from all the function arguments.

File: src/Traits/AdapterHelper.php [hunk 23]
```diff
@@ -12,25 +12,25 @@
...
-            if ($group && !$group->filterable) {
+            if ($propertyGroupOptionEntity->group && !$propertyGroupOptionEntity->group->filterable) {
                 continue;
             }
-
-            $attributes = array_merge($attributes, static::getAttributePropertyAsAttribute(
-                $propertyGroupOptionEntity,
-                $pluginConfig,
-            ));
+            $this->propertyGroupOptionEntity = $propertyGroupOptionEntity;
```
Comment (TobiasGraml11) [inline src/Traits/AdapterHelper.php:32]: Why do you need to save this in a member variable?
There should not be a lot of code changes needed, if you move the functions. The logic is exactly the same, but just shared by 2 adapters.

File: src/Export/Adapters/AdapterFactory.php [hunk 24]
```diff
@@ -153,4 +161,14 @@ public function getUserGroupsAdapter(): UserGroupsAdapter
     {
         return $this->userGroupsAdapter;
     }
+
+    public function getVariantConfigurationAdapter(): VariantConfigurationAdapter
+    {
+        return $this->variantConfigurationAdapter;
+    }
+
+    public function getOptionsAdapter(): OptionsAdapter
```
Comment (TobiasGraml11) [inline src/Export/Adapters/AdapterFactory.php:170]: Please also move this up to it's alphabetical position

File: src/Export/Adapters/AttributeAdapter.php [hunk 25]
```diff
@@ -11,14 +11,15 @@
...
+use FINDOLOGIC\Shopware6Common\Traits\AdapterHelper;
 use Symfony\Contracts\Translation\TranslatorInterface;
 use Vin\ShopwareSdk\Data\Entity\Category\CategoryCollection;
 use Vin\ShopwareSdk\Data\Entity\Product\ProductEntity;
-use Vin\ShopwareSdk\Data\Entity\PropertyGroupOption\PropertyGroupOptionCollection;
-use Vin\ShopwareSdk\Data\Entity\PropertyGroupOption\PropertyGroupOptionEntity;
 
 class AttributeAdapter
 {
+    use AdapterHelper;
```
Comment (TobiasGraml11) [inline src/Export/Adapters/AttributeAdapter.php:21]: You should be able to remove `protected PluginConfig $pluginConfig;`, as you add this helper.

File: src/Export/Adapters/AttributeAdapter.php [hunk 26]
```diff
@@ -149,68 +150,24 @@ protected function getManufacturerAttributes(ProductEntity $product): array
     }
 
     /**
+     * @deprecated tag:6.0.0 - Logic was moved to the OptionsAdapter
      * @return Attribute[]
      */
-    protected function getPropertyAttributes(ProductEntity $product): array
+    public function getOptionAttributes(ProductEntity $product): array
```
Comment (TobiasGraml11) [inline src/Export/Adapters/AttributeAdapter.php:156]: Why was this changed to public now? This will break all the extensions using it and should not be needed.

File: src/Export/Adapters/AttributeAdapter.php [hunk 27]
```diff
@@ -149,68 +150,24 @@ protected function getManufacturerAttributes(ProductEntity $product): array
...
-            if (!Utils::isEmpty($groupName) && !Utils::isEmpty($propertyGroupOptionName)) {
-                $propertyGroupAttrib = new Attribute($groupName);
-                $propertyGroupAttrib->addValue(Utils::removeControlCharacters($propertyGroupOptionName));
-
-                $attributes[] = $propertyGroupAttrib;
-            }
-        }
-
-        return $attributes;
+        return $this->getPropertyGroupOptionAttributes($product->properties, $this->pluginConfig);
```
Comment (TobiasGraml11) [inline src/Export/Adapters/AttributeAdapter.php:170]: This is how it should look like, when having a trait with the member variable.

```suggestion
        return $this->getPropertyGroupOptionAttributes($product->properties);
```

File: src/Export/Adapters/AttributeAdapter.php [hunk 28]
```diff
@@ -223,7 +180,7 @@ protected function getCustomFieldAttributes(ProductEntity $product): array
         }
 
         foreach ($productFields as $key => $value) {
-            $key = $this->getAttributeKey($key);
+            $key = $this->getAttributeKey($key, $this->pluginConfig);
```
Comment (TobiasGraml11) [inline src/Export/Adapters/AttributeAdapter.php:183]:
```suggestion
            $key = $this->getAttributeKey($key);
```

File: src/Export/Adapters/VariantConfigurationAdapter.php [hunk 29]
```diff
@@ -0,0 +1,56 @@
...
+use FINDOLOGIC\Shopware6Common\Traits\AdapterHelper;
+use Vin\ShopwareSdk\Data\Entity\Product\ProductEntity;
+use Vin\ShopwareSdk\Data\Entity\PropertyGroupOption\PropertyGroupOptionEntity;
+
+class VariantConfigurationAdapter
+{
+    use AdapterHelper;
+
+    protected PluginConfig $pluginConfig;
+
```
Comment (TobiasGraml11) [inline src/Export/Adapters/VariantConfigurationAdapter.php:19]:
```suggestion
```

File: tests/Traits/AdapterHelper.php [hunk 30]
```diff
@@ -170,4 +175,14 @@ public function getUserGroupAdapter(?CustomerGroupCollection $customerGroupColle
     {
         return new UserGroupsAdapter($this->getExportContext($customerGroupCollection));
     }
+
+    public function getVariantConfigurationAdapter($config): VariantConfigurationAdapter
+    {
+        return new VariantConfigurationAdapter($config ?? $this->getPluginConfig());
+    }
+
+    public function getOptionsAdapter($config): OptionsAdapter
```
Comment (TobiasGraml11) [inline tests/Traits/AdapterHelper.php:184]: Please keep the alphabetical order

**Review by TobiasGraml11 — CHANGES_REQUESTED** [review]

File: src/Export/Adapters/AdapterFactory.php [hunk 31]
```diff
@@ -38,6 +38,10 @@ class AdapterFactory
 
     private UserGroupsAdapter $userGroupsAdapter;
 
+    private VariantConfigurationAdapter $variantConfigurationAdapter;
+
+    private OptionsAdapter $optionsAdapter;
+
```
Comment (TobiasGraml11) [inline src/Export/Adapters/AdapterFactory.php:44]: Alphabetical order

File: src/Traits/AdapterHelper.php [hunk 32]
```diff
@@ -0,0 +1,72 @@
...
+        $attributes = [];
+
+        foreach ($collection as $propertyGroupOptionEntity) {
+            if (
+                $propertyGroupOptionEntity->group
+                &&
+                !$propertyGroupOptionEntity->group->filterable
+            ) {
+                continue;
+            }
```
Comment (TobiasGraml11) [inline src/Traits/AdapterHelper.php:31]: Please keep the code as it was before.

```suggestion
            $group = $propertyGroupOptionEntity->group;
            if ($group && !$group->filterable) {
                continue;
            }
```

File: src/Export/Adapters/ExportItemAdapter.php [hunk 33]
```diff
@@ -24,15 +25,19 @@ class ExportItemAdapter
...
     private LoggerInterface $logger;
 
+    protected PluginConfig $pluginConfig;
+
     private ?EventDispatcherInterface $eventDispatcher;
 
     public function __construct(
         AdapterFactory $adapterFactory,
         LoggerInterface $logger,
+        PluginConfig $pluginConfig,
```
Comment (TobiasGraml11) [inline src/Export/Adapters/ExportItemAdapter.php:35]: The config is not used, you can remove it

File: src/Export/Adapters/ExportItemAdapter.php [hunk 34]
```diff
@@ -157,6 +169,12 @@ public function adaptVariant(Item $item, ProductEntity $product): ?Item
             foreach ($this->adapterFactory->getShopwarePropertiesAdapter()->adapt($product) as $property) {
                 $item->addProperty($property);
             }
+
+            $optionAttributes = $this->adapterFactory->getVariantConfigurationAdapter()->getOptionAttributes($product);
+
+            foreach ($optionAttributes as $attribute) {
+                $item->addMergedAttribute($attribute);
+            }
```
Comment (TobiasGraml11) [inline src/Export/Adapters/ExportItemAdapter.php:177]:
```suggestion
            $optionAttributes = $this->adapterFactory->getVariantConfigurationAdapter()->getOptionAttributes($product);
            foreach ($optionAttributes as $attribute) {
                $item->addMergedAttribute($attribute);
            }
```

File: src/Export/Adapters/OptionsAdapter.php [hunk 35]
```diff
@@ -0,0 +1,32 @@
...
+
+    public function __construct(PluginConfig $pluginConfig)
+    {
+        $this->pluginConfig = $pluginConfig;
+    }
+
+    /**
+     * @return Attribute[]
+     */
+    public function getOptionAttributes(ProductEntity $product): array
```
Comment (TobiasGraml11) [inline src/Export/Adapters/OptionsAdapter.php:24]: Every adapter should just have one public function with the name `adapt()`, please rename this function.

Maybe it would make sense to create an `AdapterInterface` with the `adapt()` function and add `implements AdapterInterface` to all the adapters.

(`ExportItemAdapter` is an exception)

File: src/Export/Adapters/VariantConfigurationAdapter.php [hunk 36]
```diff
@@ -0,0 +1,53 @@
...
+    public function __construct(PluginConfig $pluginConfig)
+    {
+        $this->pluginConfig = $pluginConfig;
+    }
+
+    /**
+     * @return Attribute[]
+     */
+    public function getOptionAttributes(ProductEntity $product): array
```
Comment (TobiasGraml11) [inline src/Export/Adapters/VariantConfigurationAdapter.php:26]: Same as for the OptionsAdapter

File: src/Export/Adapters/VariantConfigurationAdapter.php [hunk 37]
```diff
@@ -0,0 +1,53 @@
...
+            $product->variantListingConfig['configuratorGroupConfig'] ?? [],
+            function (array $listing) {
+                return $listing['expressionForListings'];
+            },
+        );
+        $variantListingGroupId = array_map(fn ($listing) => $listing['id'], $variantlisting);
+
+        if (
+            $this->pluginConfig->getMainVariant() === MainVariant::SHOPWARE_DEFAULT &&
+            count($variantlisting)
```
Comment (TobiasGraml11) [inline src/Export/Adapters/VariantConfigurationAdapter.php:44]: Within my suggested code on Youtrack, I added `&& !product->variantListingConfig['displayParent']`.
Is this not needed, or was it forgotten?

File: tests/Export/Adapters/OptionAdapterTest.php [hunk 38]
```diff
@@ -0,0 +1,39 @@
...
+
+    public OptionsAdapter $optionsAdapter;
+    public PluginConfig $pluginConfig;
+    public function setUp(): void
+    {
+        $this->pluginConfig = $this->getPluginConfig();
+        $this->optionsAdapter = $this->getOptionsAdapter($this->pluginConfig);
+    }
+
+    public function testGetOptionAttributes(): void
```
Comment (TobiasGraml11) [inline tests/Export/Adapters/OptionAdapterTest.php:27]: There are no tests for the `VariantConfigurationAdapter`.

It's important to test the different `variantListingConfigs`.

File: tests/Traits/AttributeHelper.php (hunk 11 above)
Comment (TobiasGraml11) [inline tests/Traits/AttributeHelper.php:50]: You could move it to a `getOptions()` function, so you can make the assertion like we do for the attributes. (Check the usage of `getAttributes()`)

**Review by TobiasGraml11 — CHANGES_REQUESTED** [review]

File: src/Export/Adapters/ExportItemAdapter.php [hunk 39]
```diff
@@ -24,6 +25,8 @@ class ExportItemAdapter
 
     private LoggerInterface $logger;
 
+    protected PluginConfig $pluginConfig;
```
Comment (TobiasGraml11) [inline src/Export/Adapters/ExportItemAdapter.php:28]: Member variable is unused

File: tests/Export/Adapters/ExportItemAdapterTest.php [hunk 40]
```diff
@@ -26,13 +27,20 @@ class ExportItemAdapterTest extends TestCase
...
+    }
+
     public function testEventsAreDispatched(): void
     {
         $xmlItem = new XMLItem(Uuid::randomHex());
         $product = $this->createTestProduct();
 
         $eventDispatcherMock = $this->getEventDispatcherMock();
-        $eventDispatcherMock->expects($this->exactly(4))
+        $eventDispatcherMock->expects($this->exactly(3))
```
Comment (TobiasGraml11) [inline tests/Export/Adapters/ExportItemAdapterTest.php:43]: The amount of dispatched events should not have changes

File: src/Traits/AdapterHelper.php [hunk 41]
```diff
@@ -0,0 +1,69 @@
...
+trait AdapterHelper
+{
+    protected PluginConfig $pluginConfig;
+
+    /**
+     * @return Attribute[]
+     */
+    protected function getPropertyGroupOptionAttributes(
+        PropertyGroupOptionCollection $collection
+    ): array {
```
Comment (TobiasGraml11) [inline src/Traits/AdapterHelper.php:21]: The line length seems to be short enough

```suggestion
    protected function getPropertyGroupOptionAttributes(PropertyGroupOptionCollection $collection): array
    {
```

File: src/Traits/AdapterHelper.php [hunk 42]
```diff
@@ -0,0 +1,69 @@
...
+            $attributes = array_merge($attributes, $this->getAttributePropertyAsAttribute($propertyGroupOptionEntity));
+        }
+
+        return $attributes;
+    }
+
+    /**
+     * @return Attribute[]
+     */
+    protected function getAttributePropertyAsAttribute($propertyGroupOptionEntity): array
```
Comment (TobiasGraml11) [inline src/Traits/AdapterHelper.php:38]: Typehint is missing

File: src/Export/Adapters/ExportItemAdapter.php [hunk 43]
```diff
@@ -157,6 +169,12 @@ public function adaptVariant(Item $item, ProductEntity $product): ?Item
             foreach ($this->adapterFactory->getShopwarePropertiesAdapter()->adapt($product) as $property) {
                 $item->addProperty($property);
             }
+
+            $optionAttributes = $this->adapterFactory->getVariantConfigurationAdapter()->adapt($product);
+
+            foreach ($optionAttributes as $attribute) {
+                $item->addMergedAttribute($attribute);
+            }
```
Comment (TobiasGraml11) [inline src/Export/Adapters/ExportItemAdapter.php:177]: With the use of the `adapt` function, this should fit into the max line length

```suggestion
            foreach ($this->adapterFactory->getVariantConfigurationAdapter()->adapt($product) as $attribute) {
                $item->addMergedAttribute($attribute);
            }
```

File: tests/Export/Adapters/ExportItemAdapterTest.php [hunk 44]
```diff
@@ -62,9 +70,10 @@ public function testExceptionIsNotThrownForVariantWithNoCategories(): void
         $product = $this->createTestProduct([
             'id' => $id,
             'categories' => [],
+            'variantListingConfig' => ['displayParent' => null],
```
Comment (TobiasGraml11) [inline tests/Export/Adapters/ExportItemAdapterTest.php:73]: Probably makes sense to add this as default value, it should not make issues with the existing tests

File: tests/Export/Adapters/ExportItemAdapterTest.php [hunk 45]
```diff
@@ -26,13 +27,20 @@ class ExportItemAdapterTest extends TestCase
     use ProductHelper;
     use ServicesHelper;
 
+    public PluginConfig $pluginConfig;
```
Comment (TobiasGraml11) [inline tests/Export/Adapters/ExportItemAdapterTest.php:30]: Can be removed, once it was also removed within the `ExportItemAdapter`

File: tests/Traits/AdapterHelper.php [hunk 46]
```diff
@@ -115,6 +119,11 @@ public function getNameAdapter(): NameAdapter
         return new NameAdapter();
     }
 
+    public function getOptionsAdapter($config): OptionsAdapter
```
Comment (TobiasGraml11) [inline tests/Traits/AdapterHelper.php:122]:
```suggestion
    public function getOptionsAdapter(?PluginConfig $config = null): OptionsAdapter
```

File: tests/Traits/AdapterHelper.php [hunk 47]
```diff
@@ -170,4 +179,9 @@ public function getUserGroupAdapter(?CustomerGroupCollection $customerGroupColle
     {
         return new UserGroupsAdapter($this->getExportContext($customerGroupCollection));
     }
+
+    public function getVariantConfigurationAdapter($config): VariantConfigurationAdapter
```
Comment (TobiasGraml11) [inline tests/Traits/AdapterHelper.php:183]:
```suggestion
    public function getVariantConfigurationAdapter(?PluginConfig $config = null): VariantConfigurationAdapter
```

**Review by TobiasGraml11 — CHANGES_REQUESTED** [review]

File: tests/Export/Adapters/ExportItemAdapterTest.php (hunk 40 above)
Comment (bojand-soprex) [reply tests/Export/Adapters/ExportItemAdapterTest.php]: The amount was 3 so i hanged it to 4 after adding  'variantListingConfig' => ['displayParent' => null], in ProductHelper.php since now the exact exprected/acutall count is 4.

Comment (TobiasGraml11) [reply tests/Export/Adapters/ExportItemAdapterTest.php]: If random expected values change, please investigate it instead of changing them.

File: tests/Export/Adapters/ExportItemAdapterTest.php [hunk 48]
```diff
@@ -26,6 +26,11 @@ class ExportItemAdapterTest extends TestCase
     use ProductHelper;
     use ServicesHelper;
 
+    public function setUp(): void
+    {
+        $this->pluginConfig = $this->getPluginConfig();
```
Comment (TobiasGraml11) [inline tests/Export/Adapters/ExportItemAdapterTest.php:31]: You can remove this here and within the test, as you are only using the default one anyway.

**Review by TobiasGraml11 — CHANGES_REQUESTED** [review]

**Review by TobiasGraml11 — APPROVED** [review]
